src/System.Threading.Thread/tests/ThreadTests.cs

```diff
@@ -427,6 +429,58 @@ public static void CurrentPrincipalTest()
             });
         }
 
+        [Fact]
+        public static void CurrentPrincipalContextFlowTest()
+        {
+            ThreadTestHelpers.RunTestInBackgroundThread(async () =>
+            {
+                Thread.CurrentPrincipal = new ClaimsPrincipal();
+
+                await Task.Run(async() => {
+
+                    Assert.IsType<ClaimsPrincipal>(Thread.CurrentPrincipal);
+
+                    await Task.Run(async() => 
+                    {
+                        Assert.IsType<ClaimsPrincipal>(Thread.CurrentPrincipal);
+
+                        Thread.CurrentPrincipal = new GenericPrincipal(new GenericIdentity("name"), new string[0]);
+
+                        await Task.Run(() =>
+                        {
+                            Assert.IsType<GenericPrincipal>(Thread.CurrentPrincipal);
+                        });
+
+                        Assert.IsType<GenericPrincipal>(Thread.CurrentPrincipal);
+                    });
+
+                    Assert.IsType<ClaimsPrincipal>(Thread.CurrentPrincipal);
+                });
+
+                Assert.IsType<ClaimsPrincipal>(Thread.CurrentPrincipal);
+            });
+        }
+
+        [Fact]
+        public static void CurrentPrincipalContextFlowTest_NotFlow()
+        {
+            ThreadTestHelpers.RunTestInBackgroundThread(async () =>
+            {
+                Thread.CurrentPrincipal = new ClaimsPrincipal();
+
+                // TaskCreationOptions.LongRunning to have fresh new thread with clean execution context
+                await Task.Factory.StartNew(() => Assert.IsType<ClaimsPrincipal>(Thread.CurrentPrincipal), CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default);
+
+                ExecutionContext.SuppressFlow();
+
+                // TaskCreationOptions.LongRunning to have fresh new thread with clean execution context
+                await Task.Factory.StartNew(() => Assert.Null(Thread.CurrentPrincipal), CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default);
+
+                // defaults are null for netcorapp and GenericPrincipal for netfx
+                Assert.True(Thread.CurrentPrincipal is null || Thread.CurrentPrincipal is GenericPrincipal);
```

Comment:
Hmm, I don't know if this is going to behave the way you're hoping; I'm a little worried this may result in sporadic test failures (I could be wrong). With flow suppressed, I suspect the async local's changed callback won't be invoked, which means this thread's t_principal won't be set to null... if this continuation then ends up running on the same thread that was running previously running it, it'll end up seeing CurrentPrincipal as a ClaimsPrincipal rather than null or GenericPrincipal.

Reply:
I believe that your suspect is correct...this could be the reason of ubuntu fail, I would try with custom task factory(new thread for every await) to be sure to have clean thread. About it I've got a question, you state `if this continuation then ends up running on the same thread that was running previously running it` this mean that there isn't context/tls cleanup when a thread returns to pool?If so, isn't this dangerous?

Reply:
> this mean that there isn't context/tls cleanup when a thread returns to pool?If so, isn't this dangerous?

Thread locals are not reset for every work item on a thread pool thread.  That would lead to really bad performance in common cases, e.g. where you cache things in a ThreadStatic, as is done in ArrayPool, or StringBuilderCache, or other such places.

Reply:
understood thank's.

Reply:
```cs
private static async Task Test()
{
    Thread.CurrentPrincipal = new ClaimsPrincipal();

    ExecutionContext.SuppressFlow();

    await Task.Factory.StartNew(() => Console.WriteLine(Thread.CurrentPrincipal), CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default);

    await Task.Run(() => Console.WriteLine(Thread.CurrentPrincipal));
}
...
.NET Full C:\Windows\Microsoft.NET\Framework\v4.0.30319\mscorlib.dll
System.Security.Claims.ClaimsPrincipal
System.Security.Principal.GenericPrincipal
```
On full framework suppress flow doesn't work for `StartNew` is it expected?

Reply:
Ah good to know thanks's for the infos, I removed test from netfx because to avoid issue with tls I need a new thread, I removed also second check, it doesn't have a lot of sense we neeed only to check if context is not flowed not what is context after.